lib/bolt/config/validator.rb

```diff
@@ -0,0 +1,179 @@
+      # created during validation are concatenated into a single error
+      # message.
+      #
+      private def raise_error(location)
+        return unless @errors.any?
+
+        message = "Invalid configuration"
+        message += " at #{location}" if location
+        message += ":\n"
+        message += @errors.map { |error| "\s\s#{error}" }.join("\n")
+
+        raise Bolt::ValidationError, message
+      end
+
+      # Validate an individual value. This performs validation that is
+      # common to all values, including type validation. After validating
+      # the value's type, the value is passed off to an individual
+      # validation method for the value's type.
+      #
+      private def validate_value(value, definition)
+        return if plugin_reference?(value, definition)
+        return unless valid_type?(value, definition)
+
+        case value
+        when Hash
+          validate_hash(value, definition)
+        when Array
+          validate_array(value, definition)
+        when String
+          validate_string(value, definition)
+        when Numeric
+          validate_number(value, definition)
+        end
+      end
+
+      # Validates a hash value, logging errors for any validations that fail.
+      # This will enumerate each key-value pair in the hash and validate each
+      # value individually.
+      #
+      private def validate_hash(value, definition)
+        properties = definition[:properties] ? definition[:properties].keys : []
+
+        if definition[:required] && (definition[:required] - value.keys).any?
+          missing = definition[:required] - value.keys
+          @errors << "Value at '#{path}' is missing required keys #{missing.join(', ')}"
+        end
+
+        value.each_pair do |key, val|
+          @path.push(key)
+
+          if properties.include?(key)
+            validate_value(val, definition[:properties][key])
+          elsif definition[:additionalProperties]
+            validate_value(val, definition[:additionalProperties])
+          end
+        ensure
+          @path.pop
+        end
+      end
+
+      # Validates an array value, logging errors for any validations that fail.
+      # This will enumerate the items in the array and validate each item
+      # individually.
+      #
+      private def validate_array(value, definition)
+        if definition[:uniqueItems] && value.size != value.uniq.size
+          @errors << "Value at '#{path}' must not include duplicate elements"
+          return
+        end
+
+        return unless definition.key?(:items)
+
+        value.each_with_index do |item, index|
+          @path.push(index)
+          validate_value(item, definition[:items])
+        ensure
+          @path.pop
+        end
+      end
+
+      # Validates a string value, logging errors for any validations that fail.
+      #
+      private def validate_string(value, definition)
+        if definition.key?(:enum) && !definition[:enum].include?(value)
```

Comment:
Hm...it seems like the transport option isn't getting picked up here and I can't figure out why...:
```
Unknown transport foo found for localhost:20022
```
That's coming from `lib/bolt/inventory/target.rb#154` I think

Reply:
Ah, `transport` isn't getting added into the schema. 😬 I'll fix that.

Reply:
Oh, actually, the target-specific `transport` option is handled a bit differently, since the transport can be set in multiple locations and the transport config object only handles settings under `ssh`, `winrm`, etc. I think for now we can leave the validation here as-s: https://github.com/puppetlabs/bolt/blob/main/lib/bolt/inventory/target.rb#L153-L155

Reply:
Ok, that seems fine - it only stuck out because it's different but I don't think it matters at all for the user experience.